src/other/index.jl

```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
     x
 end
 
-Base.getindex(x::Index, idx::Symbol) = x.lookup[idx]
-Base.getindex(x::AbstractIndex, idx::Real) = Int(idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Union{Bool, Missing}}) =
-    getindex(x, collect(Missings.replace(idx, false)))
+Base.getindex(x::AbstractIndex, idx::Symbol) = x.lookup[idx]
+Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+function Base.getindex(x::AbstractIndex, idx::Real)
+    idx isa Bool && Base.depwarn("Indexing with Bool values is deprecated", :getindex)
+    idx isa Integer || Base.depwarn("Indexing with values that are not Integer is deprecated", :getindex)
+    Int(idx)
+end
+
+function Base.getindex(x::AbstractIndex, idx::AbstractVector{<:Real})
+    any(v -> v isa Bool, idx) && Base.depwarn("Indexing with Bool values is deprecated")
+    all(v -> v isa Integer, idx) || Base.depwarn("Indexing with values that are not Integer is deprecated", :getindex)
+    Vector{Int}(idx)
+end
+
 function Base.getindex(x::AbstractIndex, idx::AbstractVector{Bool})
     length(x) == length(idx) || throw(BoundsError(x, idx))
     find(idx)
 end
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing} =
-    getindex(x, collect(skipmissing(idx)))
-Base.getindex(x::AbstractIndex, idx::AbstractRange) = [idx;]
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T <: Real} = convert(Vector{Int}, idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+Base.getindex(x::AbstractIndex, idx::AbstractRange) = getindex(x, collect(idx))
+
+# catch all method handling cases when type of idx is not narrowest possible, Any in particular
+# also it handles passing missing values in idx
+function Base.getindex(x::AbstractIndex, idx::AbstractVector)
+    idxs = filter(!ismissing, idx)
+    if length(idxs) != length(idx)
+        Base.depwarn("using missing in column indexing is deprecated", :getindex)
+    end
+    length(idxs) == 0 && return Int[] # special case of empty idxs
+    # below is not very efficient, but has to be done as in Bool case we replaced missing by false
+    # after deprecation period when Missing is disallowed this will not be a problem
```

Comment:
AFAICT we previously didn't support logical indexing via `AbstractVector{Any}`. So this deprecation could be implemented via a method for `AbstractArray{<:Union{Missing,Bool}}` in deprecated.jl to keep this one clean.

Reply:
OK